### PR TITLE
feat(batcher): add abort to the BlockBuilder

### PR DESCRIPTION
**Pull request by dafnamatsry**

### CODE REVIEW

**Comment by dafnamatsry** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/1934?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1934** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/1934?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1910** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/1910?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dafnamatsry and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by alonh5 — CHANGES_REQUESTED** [review]
> Reviewed 2 of 4 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1934)* status: all files reviewed, 2 unresolved discussions (waiting on @dafnamatsry and @Yael-Starkware)

___
*[`crates/batcher/src/proposal_manager.rs` line 173 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1934#-OBPqMlC9Mzx2T8K6bTD:-OBPqMlC9Mzx2T8K6bTE:b-2j1ra9) ([raw file](https://github.com/starkware-libs/sequencer/blob/7249597a7977c42f9bff125d6e8e682050cda4ec/crates/batcher/src/proposal_manager.rs#L173)):*
> ```XML
> 
>         // Create the block builder, and a channel to allow aborting the block building task.
>         let (_abort_signal_sender, abort_signal_receiver) = tokio::sync::oneshot::channel();
> ```

Using the sender in the next PR?



_Code quote:_
```XML
_abort_signal_sender
```
___
*[`crates/batcher/src/block_builder_test.rs` line 26 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1934#-OBPqrKo7oIY_zcGOu7C:-OBPqrKo7oIY_zcGOu7D:b836jym) ([raw file](https://github.com/starkware-libs/sequencer/blob/7249597a7977c42f9bff125d6e8e682050cda4ec/crates/batcher/src/block_builder_test.rs#L26)):*
> ```XML
> use crate::transaction_provider::{MockTransactionProvider, NextTxs};
> 
> const BLOCK_GENERATION_DEADLINE_SECS: u64 = 5;
> ```

Does this make some tests unecessarily slow?



_Code quote:_
```XML
const BLOCK_GENERATION_DEADLINE_SECS: u64 = 5;
```

**Review by dafnamatsry — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1934)* status: all files reviewed, 2 unresolved discussions (waiting on @alonh5 and @Yael-Starkware)

___
*[`crates/batcher/src/block_builder_test.rs` line 26 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1934#-OBPqrKo7oIY_zcGOu7C:-OBPrOJGDSvT-YgU34Fx:b-sab1pd) ([raw file](https://github.com/starkware-libs/sequencer/blob/7249597a7977c42f9bff125d6e8e682050cda4ec/crates/batcher/src/block_builder_test.rs#L26)):*
<details><summary><i>Previously, alonh5 (Alon Haramati) wrote…</i></summary><blockquote>

Does this make some tests unecessarily slow?
</blockquote></details>

IIUC it affects only the test that checks the deadline, since we wait for it to be reached.
Do you want me to make it a  variable?
___
*[`crates/batcher/src/proposal_manager.rs` line 173 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1934#-OBPqMlC9Mzx2T8K6bTD:-OBPrMLl3Lqsij06vd9i:bb77ufg) ([raw file](https://github.com/starkware-libs/sequencer/blob/7249597a7977c42f9bff125d6e8e682050cda4ec/crates/batcher/src/proposal_manager.rs#L173)):*
<details><summary><i>Previously, alonh5 (Alon Haramati) wrote…</i></summary><blockquote>

Using the sender in the next PR?
</blockquote></details>

Yes.

**Review by alonh5 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1934)* status: all files reviewed, 1 unresolved discussion (waiting on @dafnamatsry and @Yael-Starkware)

___
*[`crates/batcher/src/block_builder_test.rs` line 26 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1934#-OBPqrKo7oIY_zcGOu7C:-OBPyYjYCyZBVhJIG9p8:b-s3qzh) ([raw file](https://github.com/starkware-libs/sequencer/blob/7249597a7977c42f9bff125d6e8e682050cda4ec/crates/batcher/src/block_builder_test.rs#L26)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

IIUC it affects only the test that checks the deadline, since we wait for it to be reached.
Do you want me to make it a  variable?
</blockquote></details>

Yes let's try that

**Review by Yael-Starkware — CHANGES_REQUESTED** [review]
> Reviewed 2 of 4 files at r1, 1 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1934)* status: all files reviewed, 5 unresolved discussions (waiting on @alonh5 and @dafnamatsry)

___
*[`crates/batcher/src/block_builder_test.rs` line 26 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1934#-OBPqrKo7oIY_zcGOu7C:-OBUQAKi-jNWwHMjndG-:b-stunk2) ([raw file](https://github.com/starkware-libs/sequencer/blob/7249597a7977c42f9bff125d6e8e682050cda4ec/crates/batcher/src/block_builder_test.rs#L26)):*
<details><summary><i>Previously, alonh5 (Alon Haramati) wrote…</i></summary><blockquote>

Yes let's try that
</blockquote></details>

it affects all tests besides block_full and stream_done.
___
*[`crates/batcher/src/block_builder_test.rs` line 335 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1934#-OBUOr5c0CfUB82WD7ec:-OBUOr5c0CfUB82WD7ed:bt59zz7) ([raw file](https://github.com/starkware-libs/sequencer/blob/7249597a7977c42f9bff125d6e8e682050cda4ec/crates/batcher/src/block_builder_test.rs#L335)):*
> ```XML
> 
> #[rstest]
> #[tokio::test]
> ```

In this test, all provided transactions are being processed, I think it is also interesting to test a case where some of the transactions are not being processed due to the abort signal.
For example, an abort signal that was sent before the build_block call.
___
*[`crates/batcher/src/block_builder_test.rs` line 339 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1934#-OBUOCW20jE5UX3Buam5:-OBUOCW20jE5UX3Buam6:undefined) ([raw file](https://github.com/starkware-libs/sequencer/blob/7249597a7977c42f9bff125d6e8e682050cda4ec/crates/batcher/src/block_builder_test.rs#L339)):*
> ```XML
>     let mock_tx_provider = mock_tx_provider_limitless_calls(1, vec![test_txs(0..3)]);
> 
>     // Expect at least one transaction chunk to be added to the block, and then abort.
> ```





_Suggestion:_
```XML
Expect one transaction chunk to be added to the block, and then abort.
```
___
*[`crates/batcher/src/block_builder_test.rs` line 343 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1934#-OBUNX2x2lZMdp3wxQYr:-OBUNX2yDykwWDLANqG1:b45h) ([raw file](https://github.com/starkware-libs/sequencer/blob/7249597a7977c42f9bff125d6e8e682050cda4ec/crates/batcher/src/block_builder_test.rs#L343)):*
> ```XML
>     mock_transaction_executor
>         .expect_add_txs_to_block()
>         .returning(|_| (0..3).map(|_| Ok(execution_info())).collect());
> ```





_Suggestion:_
```XML
    mock_transaction_executor
        .expect_add_txs_to_block().times(1)
        .returning(|_| (0..3).map(|_| Ok(execution_info())).collect());
```
___
*[`crates/batcher/src/proposal_manager.rs` line 173 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1934#-OBUP_3b9H5M4g8RdMKL:-OBUP_3b9H5M4g8RdMKM:b-jtm5pw) ([raw file](https://github.com/starkware-libs/sequencer/blob/7249597a7977c42f9bff125d6e8e682050cda4ec/crates/batcher/src/proposal_manager.rs#L173)):*
> ```XML
> 
>         // Create the block builder, and a channel to allow aborting the block building task.
>         let (_abort_signal_sender, abort_signal_receiver) = tokio::sync::oneshot::channel();
> ```

should be used.
I assume it is in a different PR?



_Code quote:_
```XML
_abort_signal_sender
```

**Review by dafnamatsry — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1934)* status: 3 of 4 files reviewed, 5 unresolved discussions (waiting on @alonh5 and @Yael-Starkware)

___
*[`crates/batcher/src/block_builder_test.rs` line 26 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1934#-OBPqrKo7oIY_zcGOu7C:-OBUg5_1-zLsh33wi4AY:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/7249597a7977c42f9bff125d6e8e682050cda4ec/crates/batcher/src/block_builder_test.rs#L26)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

it affects all tests besides block_full and stream_done.
</blockquote></details>

Done.
___
*[`crates/batcher/src/block_builder_test.rs` line 335 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1934#-OBUOr5c0CfUB82WD7ec:-OBUcP4P3KTTJx2NALnz:bw5up0x) ([raw file](https://github.com/starkware-libs/sequencer/blob/7249597a7977c42f9bff125d6e8e682050cda4ec/crates/batcher/src/block_builder_test.rs#L335)):*
Not sure I understand. I all cases, all the provided txs will be processed, since the abort only happens after we process the last chunk we have, and then we stop asking for more txs.

> For example, an abort signal that was sent before the build_block call.

I don't think this case is interesting enough to have it tested.
___
*[`crates/batcher/src/proposal_manager.rs` line 173 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1934#-OBUP_3b9H5M4g8RdMKL:-OBUUyjp7VUDg5WoJbv5:bhwmhbs) ([raw file](https://github.com/starkware-libs/sequencer/blob/7249597a7977c42f9bff125d6e8e682050cda4ec/crates/batcher/src/proposal_manager.rs#L173)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

should be used.
I assume it is in a different PR?
</blockquote></details>

Yes, in following PR.
___
*[`crates/batcher/src/block_builder_test.rs` line 339 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1934#-OBUOCW20jE5UX3Buam5:-OBUbpex-1L5n35eSq52:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/7249597a7977c42f9bff125d6e8e682050cda4ec/crates/batcher/src/block_builder_test.rs#L339)):*
> ```XML
>     let mock_tx_provider = mock_tx_provider_limitless_calls(1, vec![test_txs(0..3)]);
> 
>     // Expect at least one transaction chunk to be added to the block, and then abort.
> ```

Done.
___
*[`crates/batcher/src/block_builder_test.rs` line 343 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1934#-OBUNX2x2lZMdp3wxQYr:-OBUbpDv6FAOnFFL14r_:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/7249597a7977c42f9bff125d6e8e682050cda4ec/crates/batcher/src/block_builder_test.rs#L343)):*
> ```XML
>     mock_transaction_executor
>         .expect_add_txs_to_block()
>         .returning(|_| (0..3).map(|_| Ok(execution_info())).collect());
> ```

Done.

**Review by alonh5 — APPROVED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1934)* status: all files reviewed, 4 unresolved discussions (waiting on @Yael-Starkware)

**Review by Yael-Starkware — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1934)* status: all files reviewed, 1 unresolved discussion (waiting on @dafnamatsry)

___
*[`crates/batcher/src/block_builder_test.rs` line 335 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1934#-OBUOr5c0CfUB82WD7ec:-OBVRBWZ7Bku3r7nPM9y:bn6mewp) ([raw file](https://github.com/starkware-libs/sequencer/blob/7249597a7977c42f9bff125d6e8e682050cda4ec/crates/batcher/src/block_builder_test.rs#L335)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

Not sure I understand. I all cases, all the provided txs will be processed, since the abort only happens after we process the last chunk we have, and then we stop asking for more txs.

> For example, an abort signal that was sent before the build_block call.

I don't think this case is interesting enough to have it tested.
</blockquote></details>

let me rephrase, add a test case where some of the transactions remain in the tx_provider and not being asked for by the block_builder.
if someone inserts a bug by moving the abort signal check to the end of the build_block function, the function will return an abort error eventually, but won't stop asking for transactions till deadline.

**Review by dafnamatsry — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1934)* status: 0 of 8 files reviewed, 1 unresolved discussion (waiting on @alonh5, @graphite-app[bot], and @Yael-Starkware)

___
*[`crates/batcher/src/block_builder_test.rs` line 335 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1934#-OBUOr5c0CfUB82WD7ec:-OBZoS10CqpajJUnIkay:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/7249597a7977c42f9bff125d6e8e682050cda4ec/crates/batcher/src/block_builder_test.rs#L335)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

let me rephrase, add a test case where some of the transactions remain in the tx_provider and not being asked for by the block_builder.
if someone inserts a bug by moving the abort signal check to the end of the build_block function, the function will return an abort error eventually, but won't stop asking for transactions till deadline.
</blockquote></details>

Done.

**Review by alonh5 — APPROVED** [review]
> Reviewed 4 of 8 files at r4.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1934)* status: all files reviewed, 1 unresolved discussion (waiting on @Yael-Starkware)

**Review by alonh5 — APPROVED** [review]
> Reviewed 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1934)* status: all files reviewed, 1 unresolved discussion (waiting on @Yael-Starkware)

**Review by alonh5 — APPROVED** [review]
> Reviewed 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1934)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @dafnamatsry)
